File: 2019-The Score of Students Solving Math Expression.py
class Solution(object):
    def scoreOfStudents(self, s, answers):
        """
        :type s: str
        :type answers: List[int]
        :rtype: int
        """
        correct = eval(s)
        
        candidate = set()
        memo = dict()
        
        def cal(start, end):
            if (start, end) in memo:
                return memo[(start, end)]
            if start == end:
                return [int(s[start])]
            
            res = set()
            
            for i in range(start, end+1):
                if s[i] in ("+", "-", "*"):
                    left = cal(start, i-1)
                    right = cal(i+1, end)
                    for l in left:
                        for r in right:
                            if s[i] == "+" and l + r <= 1000:
                                res.add(l+r)
                            elif s[i] == "*" and l*r <= 1000:
                                res.add(l*r)
            memo[(start, end)] = res
            return res
        
        candidates = cal(0, len(s)-1)

        ans = 0
        for answer in answers:
            if answer == correct:
                ans+=5
            elif answer in candidates:
                ans+=2
                
        return ans

File: JavaScript/2019-The Score of Students Solving Math Expression.js
/**
 * @param {string} s
 * @param {number[]} answers
 * @return {number}
 */
var scoreOfStudents = function(s, answers) {
    let correct = 0;
    for (let toAdd of s.split("+")){
        let multiplied = 1
        for (let toMulti of toAdd.split("*")){
            multiplied *= parseInt(toMulti);
        }
        correct += multiplied;
    }

    const operator = {
        "+": ((a,b)=>a+b),
        "*": ((a,b)=>a*b)
    } 
    
    dp = new Array(30).fill().map(()=>new Array(30));
    
    let cal = (start, end) =>{
        if (start === end){
            return [parseInt(s[start])];
        }
        if (dp[start][end]){
            return dp[start][end];
        }
        let res = new Set();
        for (let i=start; i<=end; i++){
            const current = s[i];
            if (isNaN(current)){
                let left = cal(start, i-1), right = cal(i+1, end);
                for (let l of left){
                    for (let r of right){
                        let z = operator[current](l, r);
                        if (z <= 1000) res.add(operator[current](l, r));
                    }
                }
            }
        }
        dp[start][end] = res;
        return res;
    }
    
    other = cal(0, s.length-1);
    let ans = 0;
    
    for (let answer of answers){
        if (answer === correct){
            ans+=5;
        }else if(other.has(answer)){
            ans+=2;
        }
    }
    return ans;
};